mwan3: fix some tunnels assigned the wrong mark

anyuta1166 opened the pull request:
The mark of underlying tunnel connection is assigned to all incoming packets inside tunnel. This breaks mwan3 routing. Attempt to fix this by clearing the mark in incoming packets. Do not touch outgoing packets to make sure that tracking and "mwan3 use" command works as expected.

Maintainer: @feckert
Compile tested: x86-64, OpenWrt master
Run tested: x86-64, OpenWrt master

Description:

There is an issue with some tunnels. The mark of underlying tunnel connection is assigned to all incoming packets inside tunnel. This breaks the routing, as replies to these packets may be sent to the wrong interface.
This applies, for example, to ipv6 tunnels (#14332 #18481), ipsec tunnels (#19607) and so on.
Fix this by resetting the mark in incoming packets.
This patch does not touch outgoing packets and doesn't break mwan3 wrapper library.

Fixes #14332 #18481 #19607

anyuta1166 committed:
mwan3: fix some tunnels assigned the wrong mark

The mark of the underlying tunnel connection is assigned to all
incoming packets inside the tunnel. This breaks mwan3 routing.

Fix this by clearing the mark in incoming packets. Do not touch
outgoing packets to make sure that tracking and the "mwan3 use"
command works as expected.

Signed-off-by: Anna Tikhomirova <vamp@vampik.ru>

anyuta1166 commented on net/mwan3/files/lib/mwan3/mwan3.sh:
```diff
@@ -338,6 +338,16 @@ mwan3_set_general_iptables()
+			mwan3_push_update -N mwan3_pre
+			mwan3_push_update -A mwan3_pre \
+					  -m mark ! --mark "0x0/$MMX_MASK" \
+					  -j MARK --set-xmark "0x0/$MMX_MASK"
```
mwan3_hook applies to both PREROUTING and OUTPUT. We need to reset the mark in PREROUTING only, that's why I created a new chain. If we add a rule that resets the mark to mwan3_hook at the first position, it will apply to OUTPUT and break mwan3 tracking and mwan3 use command.

Regarding that you don't seem to be affected, although you also have an 6in4 tunnel. The issue is that:
1) First outgoing packet gets interface mark (for example, 0x200) and is correctly routed **via mwan3**
2) Reply packet is already marked 0x3f00 before reaching mwan3_hook, and mwan3_hook saves this mark and the whole connection gets mark 0x3f00
3) Second and all subsequent outgoing packets gets mark 0x3f00 (restored from connection mark by mwan3_hook) and goes out **via default routing table** 
You may not even notice that if you have only one IPv6 connection. But the things break when you have multiple connections.
Example: I have wan6 (native ipv6) and wanb6 (6in4). In the default routing table, wan6 has metric 1 and wanb6 has metric 2. And I have a mwan3 rule that routes outgoing connections to the specific resource via wanb6.
First outgoing packet to this resource is correctly routed **via mwan3** and goes out from wanb6.
Second and all subsequent outgoing packets goes out **via default routing table** via the wrong interface (wan6), resulting in the packet being dropped.

anyuta1166 replied:
The same applies to ipv4 ipsec tunnels - second and all subsequent outgoing packets goes out via default routing table via the main interface instead of being routed by mwan3 via the ipsec interface.

patrakov replied:
The test for `! --mark "0x0/$MMX_MASK"` is redundant. "If it is not zero, reset it to zero" and "reset it to zero no matter what" are 100% equivalent instructions.

anyuta1166 replied:
> The inner packet passes through the tables first, getting the right mark for the tunnel interface. Then it gets encapsulated, and the outer packet passes through the tables again, preserving the inner packet's mark. As the mark is already set, all iptables rules will be skipped. When it comes to making a routing decision, the right routing table won't be selected (because the mark points to the table for the tunnel interface, not for the WAN netdev). In the lucky case, it will be handled by the main table, in the unlucky case (ipv4-in-ipv4?) it may fail to be routed.

You're right. There is an issue with egress too. It works exactly as you described.
But I don't know what to do with that. If we clear the mark for outgoing packets, routing will work as expected, but it will break mwan3 tracking and 'mwan3 use' command which relies on the pre-set mark.

anyuta1166 replied:
Probably the best way would be to solve the root of the issue, i.e. to patch the linux kernel to remove retaining the packet mark after de/encapsulation. I think retaining the mark is ridiculous and not technically correct, because inner packet is not the same packet as the tunnel packet and vice versa, and they belong to the different connections, and this behavior breaks routing based on marks,
However, I'm afraid that I don't have enough knowledge to patch the linux kernel.
I believe this issue affects all kernel-based tunnels - 6in4, l2tp, ipsec, wireguard etc.

anyuta1166 replied:
Thanks for the suggestion, I've updated the PR. Made some tests and it seems that if we clear the mark in POSTROUTING, it doesn't propagate to the tunnel connection itself. 'mwan3 use' works fine too.

anyuta1166 committed:
mwan3: remove redundant check

The previous rule already handles marked packets, so the extra
condition is not needed.

Signed-off-by: Anna Tikhomirova <vamp@vampik.ru>

anyuta1166 committed:
mwan3: fix mark for outgoing connections inside tunnel

The mark of outgoing connections propagates to the tunnel
connection itself, which may break routing.

Fix this by resetting the mark of outgoing packets after the
routing decision is made.

Suggested-by: Maxim Mikityanskiy <maxtram95@gmail.com>
Signed-off-by: Anna Tikhomirova <vamp@vampik.ru>

gentoo-root reviewed (approved):
Looks good to me (I don't have a setup with two tunnels to test it, though).

mwan3_pre and mwan3_post can be merged, both chains are identical.

openwrtdiy commented:
After pressing the above repair, there are a lot of errors in the system log on my side!
https://github.com/openwrtdiy/packages/commit/798cc5fe0e4ce55ca0d24ad2c79e351aff3990e1

Architecture Intel(R) Xeon(R) Platinum 8168 CPU @ 2.70GHz
Target platform x86/64
Firmware version OpenWrt 23.05-SNAPSHOT r23396-1076526eeb / LuCI pppoe-23.05 branch git-23.220.09853-cf89ef8
Kernel version 5.15.126
mwan3 version number: 2.11.7


**The contents of some system logs are as follows:**

Fri Aug 18 13:37:08 2023 user.warn mwan3rtmon[11843]: failed: 'ip -4 route replace table 10 10.67.1.72 dev ppp327 proto kernel scope link src 10.0.0.1 ' - error: Cannot find device "ppp3 27"

Fri Aug 18 13:37:09 2023 user.warn mwan3rtmon[11843]: failed: 'ip -4 route replace table 2 10.67.5.43 dev if6950 proto kernel scope link src 10.0.0.1 ' - error: RTNETLINK answers: No such device

Fri Aug 18 13:37:09 2023 user.warn mwan3rtmon[11843]: failed: 'ip -4 route replace table 3 10.67.5.43 dev if6950 proto kernel scope link src 10.0.0.1 ' - error: RTNETLINK answers: No such device

Fri Aug 18 13:37:09 2023 user.warn mwan3rtmon[11843]: failed: 'ip -4 route replace table 4 10.67.5.43 dev if6950 proto kernel scope link src 10.0.0.1 ' - error: RTNETLINK answers: No such device

Fri Aug 18 13:37:10 2023 user.warn mwan3rtmon[11843]: failed: 'ip -4 route replace table 8 10.67.5.43 dev if6950 proto kernel scope link src 10.0.0.1 ' - error: RTNETLINK answers: No such device

Fri Aug 18 13:37:10 2023 user.warn mwan3rtmon[11843]: failed: 'ip -4 route replace table 9 10.67.5.43 dev if6950 proto kernel scope link src 10.0.0.1 ' - error: RTNETLINK answers: No such device

Fri Aug 18 13:37:10 2023 user.warn mwan3rtmon[11843]: failed: 'ip -4 route replace table 10 10.67.5.43 dev if6950 proto kernel scope link src 10.0.0.1 ' - error: RTNETLINK answers: No such device

Fri Aug 18 13:37:11 2023 user.warn mwan3rtmon[11843]: failed: 'ip -4 route replace table 2 172.31.26.9 dev if6895 proto kernel scope link src 10.0.0.1 ' - error: RTNETLINK answers: No such device

Fri Aug 18 13:37:12 2023 user.warn mwan3rtmon[11843]: failed: 'ip -4 route replace table 3 172.31.26.9 dev if6895 proto kernel scope link src 10.0.0.1 ' - error: RTNETLINK answers: No such device

Fri Aug 18 13:37:12 2023 user.warn mwan3rtmon[11843]: failed: 'ip -4 route replace table 4 172.31.26.9 dev if6895 proto kernel scope link src 10.0.0.1 ' - error: RTNETLINK answers: No such device

Fri Aug 18 13:37:12 2023 user.warn mwan3rtmon[11843]: failed: 'ip -4 route replace table 8 172.31.26.9 dev if6895 proto kernel scope link src 10.0.0.1 ' - error: RTNETLINK answers: No such device

Fri Aug 18 13:37:12 2023 user.warn mwan3rtmon[11843]: failed: 'ip -4 route replace table 9 172.31.26.9 dev if6895 proto kernel scope link src 10.0.0.1 ' - error: RTNETLINK answers: No such device

Fri Aug 18 13:37:13 2023 user.warn mwan3rtmon[11843]: failed: 'ip -4 route replace table 10 172.31.26.9 dev if6895 proto kernel scope link src 10.0.0.1 ' - error: RTNETLINK answers : No such device

anyuta1166 commented:
> After pressing the above repair, there are a lot of errors in the system log on my side!
Your issue is not related to this PR. This PR just adds a couple of iptables rules, it doesn't touch a routing table in any way.

jamesmacwhite commented:
@feckert Hi. Will this get merged? Seems like it solves a long standing issue?

Lu5ck reviewed (approved):
Tested it on 24.10.1. Applied the patch by manually editing the mentioned files. Confirm working for my FOU tunnel, fixed `ping: sendto: Operation not permitted` bug

anyuta1166 committed:
mwan3: bump PKG_VERSION to 2.12.2

Update package version after changing mwan3 scripts.

Signed-off-by: Anna Tikhomirova <vamp@vampik.ru>
